### PR TITLE
Remove zio-nio from ZMXCmlient #100

### PR DESCRIPTION
**Pull request by ahoy-jon**

### CODE REVIEW

File: src/main/scala/zio/zmx/diagnostics/ZMXClient.scala [hunk 1]
```diff
@@ -17,63 +17,70 @@
...
-        hasRemaining <- buffer.hasRemaining
-        result       <- if (hasRemaining)
-                          buffer
-                            .withJavaBuffer(buf => ZIO.succeed(acc :+ StandardCharsets.UTF_8.decode(buf).toString))
-                            .flatMap(drainBuffer(_, buffer))
-                        else ZIO.succeed(acc)
-      } yield result
+  private def nioRequestResponse(req: String): Task[String] = {
+    @tailrec
+    def drainBuffer$(acc: Chunk[String], buffer: ByteBuffer): Chunk[String] =
```
Comment (lkff) [inline src/main/scala/zio/zmx/diagnostics/ZMXClient.scala:56]: 1. Old code is back again (`(req: Chunk[Byte])` <-> `(req: String)`) etc. 😉 
2. Why `$` in the `drainBuffer$` method name?

File: src/main/scala/zio/zmx/diagnostics/ZMXClient.scala [hunk 2]
```diff
@@ -17,63 +17,70 @@
...
+      if (buffer.hasRemaining) {
+        drainBuffer$(acc :+ StandardCharsets.UTF_8.decode(buffer).toString, buffer)
+      } else acc
```
Comment (lkff) [inline src/main/scala/zio/zmx/diagnostics/ZMXClient.scala:59]: I think we should keep operating on bytes until the very end.
Here, I guess you can end up with malformed Unicode if there was non-ASCII encoded character and boundary of two buffers ended up splitting those code points.

**Review by lkff — COMMENTED** [review]
Few tips/questions :v:

File: src/main/scala/zio/zmx/diagnostics/ZMXClient.scala [hunk 3]
```diff
@@ -17,63 +17,70 @@
...
 }
 
 class ZMXClient(config: ZMXConfig) {
 
-  def sendCommand(args: Chunk[String]): ZIO[Console, Exception, String] =
-    nioRequestResponse(ZMXClient.generateRespCommand(args)).refineToOrDie[Exception]
+  def sendCommand(args: List[String]): ZIO[Console, Exception, String] = {
+    val sending: String = ZMXClient.generateRespCommand(args)
+    nioRequestResponse(sending).refineToOrDie[Exception]
+  }
```
[the remark this answers is not in the document]
Comment (ahoy-jon) [reply src/main/scala/zio/zmx/diagnostics/ZMXClient.scala:53]: Ok, I will redo the change from zio:master

File: src/main/scala/zio/zmx/diagnostics/ZMXClient.scala (hunk 1 above)
Comment (ahoy-jon) [reply src/main/scala/zio/zmx/diagnostics/ZMXClient.scala:56]: 1. Code is from yesterday, I will check what happened.
2. `drainBuffer$` is impure, it was a way to denote it (maybe not the best)

File: src/main/scala/zio/zmx/diagnostics/ZMXClient.scala (hunk 2 above)
Comment (ahoy-jon) [reply src/main/scala/zio/zmx/diagnostics/ZMXClient.scala:59]: Ok, 👌, not sure it's in the scope of the issue, I wanted to stay close to the original code and remove the dependency.

**Comment by ahoy-jon** [comment]
Sorry for the quality of the last patch, I was losing track of the work. 

@lkff you can review the new patch, I removed `drainBuffer$`.

For the `$`, I think it would be a good pattern, like we do in Clojure with `!`, to tell the code is impure (on purpose). And it should not be exposed outside the local scope.

**Comment by ahoy-jon** [comment]
Sorry for the quality of the last patch, I was losing track of the work. 

@lkff you can review the new patch, I removed `drainBuffer$`.

For the `$`, I think it would be a good pattern, like we do in Clojure with `!`, to tell the code is impure (on purpose). And it should not be exposed outside the local scope.

I will try to change the use of UTF-8 Strings internally.

File: src/main/scala/zio/zmx/diagnostics/ZMXClient.scala (hunk 2 above)
Comment (lkff) [reply src/main/scala/zio/zmx/diagnostics/ZMXClient.scala]: Sure, it can be revisited later. Just thinking out loud 😄

**Comment by lkff** [comment]
@ahoy-jon no worries! Indeed many merges were happening in the meantime.

Looks good to me, although I'm just a random contributor, one of the ZIO members will do proper review, I just wanted to help a bit 😀

> For the $, I think it would be a good pattern, like we do in Clojure with !, to tell the code is impure (on purpose). And it should not be exposed outside the local scope.

Clojure's cool! Haven't seen this pattern much in Scala tho.

> I will try to change the use of UTF-8 Strings internally.

I guess this can be done in another PR, keeping the old flow for now.

Cheers!

**Comment by ahoy-jon** [comment]
@lkff just moved the UTF-8 decoding to the end.

**Review by adamgfraser — APPROVED** [review]
Looks great! We could potentially consolidate a couple of the operations into a single effect (e.g. in `drainChannel`) but we can always do that later as an optimization if it improves performance.

**Pull request merged**